handler.go

```diff
@@ -150,41 +133,113 @@ func (h *Handler) makeAuthenticationHandler(fn func(http.ResponseWriter, *http.R
+	// Until an admin-level user exists on the cluster one, and only one operation
+	// is permitted without authentication -- the creation of the first user.
+	// This user is also automatically granted admin-level access.
+	if h.AuthenticationEnabled && len(h.server.Users()) > 0 {
+		if len(query.Statements) != 1 {
+			_, ok := query.Statements[0].(*influxql.CreateUserStatement)
+			if !ok {
+				h.error(w, "initial admin user does not exist", http.StatusUnauthorized)
+				return
+			}
 		}
-	*/
+	}
 
-	// Parse the time precision from the query params.
-	/*
-		precision, err := parseTimePrecision(urlQry.Get("time_precision"))
-		if err != nil {
-			h.error(w, err.Error(), http.StatusBadRequest)
-			return
-		}
-	*/
+	for _, s := range query.Statements {
+		switch c := s.(type) {
+		case *influxql.CreateDatabaseStatement:
+			err = h.server.CreateDatabase(c.Name)
+			if err != nil {
+				if err == ErrDatabaseExists {
+					h.error(w, "database exists", http.StatusConflict)
+				} else {
+					h.error(w, "error executing query: "+err.Error(), http.StatusInternalServerError)
+				}
+			} else {
+				w.WriteHeader(http.StatusCreated)
+			}
+		case *influxql.DropDatabaseStatement:
+			err = h.server.DeleteDatabase(c.Name)
+			if err != nil {
+				if err == ErrDatabaseNotFound {
+					h.error(w, "database not found", http.StatusInternalServerError)
+				} else {
+					h.error(w, "error executing query: "+err.Error(), http.StatusInternalServerError)
+				}
+			} else {
+				w.WriteHeader(http.StatusNoContent)
+			}
+		case *influxql.ListDatabasesStatement:
+			databases := h.server.Databases()
+			w.Header().Add("content-type", "application/json")
+			_ = json.NewEncoder(w).Encode(&databases)
+
+		case *influxql.CreateUserStatement:
+			continue
+		case *influxql.DropUserStatement:
+			continue
+
+		case *influxql.SelectStatement:
+			continue
+		case *influxql.DropSeriesStatement:
+			continue
+		case *influxql.ListSeriesStatement:
+			continue
+		case *influxql.ListMeasurementsStatement:
+			continue
+		case *influxql.ListTagKeysStatement:
+			continue
+		case *influxql.ListTagValuesStatement:
+			continue
+		case *influxql.ListFieldKeysStatement:
+			continue
+		case *influxql.ListFieldValuesStatement:
+			continue
+
+		case *influxql.GrantStatement:
+			continue
+		case *influxql.RevokeStatement:
+			continue
+
+		case *influxql.CreateRetentionPolicyStatement:
+			rp := NewRetentionPolicy(c.Name)
+			rp.Duration = c.Duration
+			rp.ReplicaN = uint32(c.Replication)
+			if err = h.server.CreateRetentionPolicy(c.Database, rp); err != nil {
+				h.error(w, "error executing query: "+err.Error(), http.StatusInternalServerError)
+			} else {
+				w.WriteHeader(http.StatusCreated)
+			}
+		case *influxql.AlterRetentionPolicyStatement:
+			rp := NewRetentionPolicy(c.Name)
+			rp.Duration = *c.Duration // Why is this a pointer, and the next?
+			rp.ReplicaN = uint32(*c.Replication)
```

Comment:
They're pointers because they're optional in the statement. If they're not specified then we should used the defaults.